### PR TITLE
feat: email verification

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?
Implements email verification controlled by a feature flag

Loom: https://www.loom.com/share/0b0b9302a0e8499fad1d6444b013337c

**Environment**: Staging(main branch) / Production

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- [x] Login as ADMIN
- [x] Enable Feature flag
- [x] Verify your account
- [x] create a new account
- [x] verify flow works as expected

### CODE REVIEW

File: apps/web/pages/api/auth/verify-email.ts
Comment (sean-brydon) [inline apps/web/pages/api/auth/verify-email.ts:19]: Find the token that is of type verify account

File: apps/web/pages/api/auth/verify-email.ts
Comment (sean-brydon) [inline apps/web/pages/api/auth/verify-email.ts:36]: If you have already completed onboarding we redirect to event types

Makes the flow much nicer as you dont get a flash of wrong content

File: apps/web/pages/auth/verify-email.tsx
Comment (sean-brydon) [inline apps/web/pages/auth/verify-email.tsx:26]: If you navigate back to this tab after verifying your account we redirect to straight to getting started

File: packages/emails/templates/account-verify-email.ts
Comment (sean-brydon) [inline packages/emails/templates/account-verify-email.ts:17]:
```suggestion
```

File: apps/web/pages/signup.tsx
Comment (sean-brydon) [inline apps/web/pages/signup.tsx:174]: Putting this here was a nicer approach than waiting for it to load on the client - it didnt always seem to load in time and disabling the button was weird if you didnt actually have it enabled 

@zomars

File: packages/prisma/migrations/20230523101834_email_verification_feature_flag/migration.sql
Comment (sean-brydon) [inline packages/prisma/migrations/20230523101834_email_verification_feature_flag/migration.sql:9]: Enable by default

File: apps/web/pages/api/auth/signup.ts
Comment (hariombalhara) [inline apps/web/pages/api/auth/signup.ts:21]: nit: Might be a good idea to introduce zod schema for this.

Comment (sean-brydon) [reply apps/web/pages/api/auth/signup.ts:21]: I've actually done this in my signup migration branch - i can bring it over for this PR if youd like? :P

File: packages/features/auth/lib/verifyEmail.ts
Comment (hariombalhara) [inline packages/features/auth/lib/verifyEmail.ts:21]:
```suggestion
  const token = randomBytes(32).toString("hex");
```

File: packages/features/auth/lib/verifyEmail.ts
Comment (hariombalhara) [inline packages/features/auth/lib/verifyEmail.ts:32]: nit: Better to use logger here

File: packages/features/auth/lib/verifyEmail.ts
Comment (hariombalhara) [inline packages/features/auth/lib/verifyEmail.ts:41]: setHours doesn't seem to have 24 as a valid value. [The range is from 0-23.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/setHours#:~:text=An%20integer%20between%200%20and%2023%20representing%20the%20hours.)

File: apps/web/pages/api/auth/verify-email.ts
Comment (hariombalhara) [inline apps/web/pages/api/auth/verify-email.ts:17]: Why do we need to use token type here? Isn't token unique ?

File: apps/web/pages/api/auth/verify-email.ts
Comment (hariombalhara) [inline apps/web/pages/api/auth/verify-email.ts:25]: What if the token has expired?

File: apps/web/pages/api/auth/verify-email.ts
Comment (sean-brydon) [reply apps/web/pages/api/auth/verify-email.ts:17]: Good point tho - it is unique on db level

File: apps/web/pages/signup.tsx
Comment (hariombalhara) [inline apps/web/pages/signup.tsx:171]: **Unrelated to this PR** but I had an observation.

With this approach there is no type safety with the slugs. What if we move the features list to a file? It would be faster as well no matter we do it on client or server.

Also, we might be able to tree shake the code as well as at compile time we know the status of the feature flag.

To allow the features list to be extended by self hosters, we can have 2 configs.
 features.base.json -> Commited in git and managed by us
 features.extended.json -> Not committed in git by us but if it exists we merge it over features.base.json
 
 cc @zomars

Comment (sean-brydon) [reply apps/web/pages/signup.tsx:171]: We're actually meant to user the useFlagsMap hook - however, this for some reason was always returning null on the signup page (works everywhere else tho)

File: apps/web/pages/signup.tsx
Comment (hariombalhara) [inline apps/web/pages/signup.tsx:182]: Screenshot 2023-05-30 at 4 37 11 PM
I see some old verificationToken related code in getServerSideProps. Is that still valid?

Comment (sean-brydon) [reply apps/web/pages/signup.tsx:182]: I believe this is for team invites off the top of my head

Comment (sean-brydon) [reply apps/web/pages/signup.tsx:182]: Yeah it is ^

File: packages/features/auth/lib/verifyEmail.ts
Comment (hariombalhara) [inline packages/features/auth/lib/verifyEmail.ts:52]: Might be better to send 429 status code here. This would cause a 500

Comment (sean-brydon) [reply packages/features/auth/lib/verifyEmail.ts:52]: Sure happy with that - we do this elsewhere too tho 🤔

File: apps/web/pages/signup.tsx
Comment (hariombalhara) [reply apps/web/pages/signup.tsx:171]: `useFlagsMap` is client only(requires another request). I like this approach better where you can do it on server.
I just noticed that there is typesafety through flags/config.ts, we can introduce a helper method with typesafety later.

File: apps/web/pages/api/auth/verify-email.ts
Comment (sean-brydon) [reply apps/web/pages/api/auth/verify-email.ts:17]: We might honestly not need it - i dont think i realised token was unique

File: packages/trpc/server/routers/viewer/auth/resendVerifyEmail.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/auth/resendVerifyEmail.handler.ts:14]:
```suggestion
    username: ctx.user.username ?? undefined,
```
or maybe make ResendEmailOptions["ctx"]["user"] optional

File: packages/trpc/server/routers/viewer/auth/resendVerifyEmail.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/auth/resendVerifyEmail.handler.ts:12]: nit: Might be a good idea to check if email is already verified or not.

File: packages/trpc/server/routers/loggedInViewer/shouldVerifyEmail.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/loggedInViewer/shouldVerifyEmail.handler.ts:17]:
```suggestion
    isVerified: isVerified || !isCalProvider,
```
can be simplified like this.

File: apps/web/pages/api/auth/verify-email.ts
Comment (zomars) [inline apps/web/pages/api/auth/verify-email.ts:22]:
```suggestion
    return res.status(401).json({ message: "No token found" });
```

File: apps/web/pages/api/auth/verify-email.ts
Comment (zomars) [inline apps/web/pages/api/auth/verify-email.ts:26]: Without returns the rest of the code still runs, which may not be a desirable behavior.

```suggestion
    return res.status(401).json({ message: "Token expired" });
```

File: apps/web/pages/api/auth/verify-email.ts
Comment (zomars) [inline apps/web/pages/api/auth/verify-email.ts:46]: Can we login the user here? Should we use the token as a One Time Password in case it opens up in another browser (as demoed in your loom)?

File: apps/web/pages/auth/verify-email.tsx
Comment (zomars) [reply apps/web/pages/auth/verify-email.tsx:26]: Can we just return a new view that says. "Email is already verified"? Maybe add an CTA that takes you back to the dashboard.

File: apps/web/pages/signup.tsx
Comment (zomars) [reply apps/web/pages/signup.tsx:174]: Ill Allow It Spanish GIF

File: packages/features/ee/common/components/LicenseRequired.tsx
Comment (zomars) [inline packages/features/ee/common/components/LicenseRequired.tsx:24]:
```suggestion
```

File: packages/prisma/schema.prisma
Comment (zomars) [reply packages/prisma/schema.prisma:284]: We aren't really changing the file here. So we can just restore it

`git checkout origin/main packages/prisma/schema.prisma`

**Review by zomars — COMMENTED** [review]
So far so good @sean-brydon Still need to tests further before merging. Left some comments as well. 🙏🏽

**Review by hariombalhara — DISMISSED** [review]
LGTM! Great work @sean-brydon 
There is just one comment that's left addressed https://github.com/calcom/cal.com/pull/9081/files#r1210204910

Tests done.
- With feature flag enabled(Default)
	- Created a new account
		- Landing on Email has been sent page.
		- Got an email. 
		- Able to reach onboarding page after clicking that link(In the browser I did signup on)
		- One onboarding is complete, I din't see the banner to verify your email. All good !!
	- Created a new account
		- Landing on Email has been sent page.
		- Retried verification email multiple times. Got multiple emails
		- Used one of the email links and flow worked like above
	- Created a new account
		- Didn't click on verification link email
		- Directly went to localhost:3000 and was taken to onboarding screen
		- After onboarding saw a banner in the app to verify my email
		- Clicked on it and received email again. Clicked on Email and the banner got removed.
	- Tried to reuse a token and it shows a Not Found error.
- Without feature flag enabled(Disabled manually)
	- I was able to signup and then went straight to onboarding after completion of which I didn't see any banner.

Blocking merge because CI is failing. Once it passes it has an approval from my side.

File: apps/web/pages/auth/verify-email.tsx
Comment (sean-brydon) [reply apps/web/pages/auth/verify-email.tsx:26]: I dont think it was clear what i meant by this.

if you switch back to this tab after you have verified your email. It navigates you to getting-started. Meaning you dont have to log in or hard refresh

File: apps/web/pages/api/auth/verify-email.ts
Comment (sean-brydon) [reply apps/web/pages/api/auth/verify-email.ts:46]: Can this be tackled in a follow up? 

You're probably really unlikely to open the email in a different browser from the one you intend on using to log in with.

Comment (zomars) [reply apps/web/pages/api/auth/verify-email.ts:46]: > Can this be tackled in a follow up?

I would insist in this same PR. But feel free to put in an issue so we can follow up.

File: apps/web/pages/signup.tsx
Comment (zomars) [reply apps/web/pages/signup.tsx:174]: I just realized that we can achieve the same effect by just doing:

```
await ssr.viewer.features.map.prefetch();
```

Will test it out and update if it works

File: apps/web/pages/api/auth/verify-email.ts
Comment (zomars) [reply apps/web/pages/api/auth/verify-email.ts]: @sean-brydon why was my suggestion ignored? We do need a return here. If there isn't the rest of the code will still run after the response which WILL cause unintended effects.

File: apps/web/pages/signup.tsx
Comment (zomars) [inline apps/web/pages/signup.tsx:35]: With #9374 this is now always defined on first render.

File: apps/web/server/lib/ssr.ts
Comment (zomars) [inline apps/web/server/lib/ssr.ts:30]: Same code as #9374 just need to merge that one first.

File: packages/features/auth/lib/verifyEmail.ts
Comment (zomars) [inline packages/features/auth/lib/verifyEmail.ts:32]: I'd rather use the getFeatureFlagMap on server-side instead of calling individual features.

**Review by zomars — APPROVED** [review]
LGTM

**Comment by shivamklr** [comment]
@sean-brydon If it is the wrong token, should we not show a proper page? Right now, it only shows the json body. Let me know your thoughts.

**Pull request merged**
